### DISCUSSION TITLE
Metadata file specification for bookmarks/highlights and programmatically introducing highlight notes

### DISCUSSION THREAD

**Discussion by lucasmyers97** [koreader/koreader#12138] [Q&A]
# Context

I have two e-books, one in Spanish and one in English. I would like to (programmatically) attach metadata to the Spanish e-book which consists of highlights over each sentence, and a note containing the English translation. I'm fine to do this in python, but I need to be able to correctly write to the `metadata.epub.lua` (which I am having trouble understanding).

# Concise questions

1. Is there somewhere which explains the "bookmarks" and "highlight" table formats of the `metadata.epub.lua` file? 
2. How do these tables specify the positions of highlights/bookmarks, and how does that relate to the epub `xhtml` files?
3. Will I run into problems if I generate my own `metadata.epub.lua` and copy that into the book's folder? 

# Attempt to understand the format + confusion

In the `metadata.epub.lua` , there's a `"bookmarks"` table whose entries look as follows:
``` lua
[1] = {
    ["chapter"] = "Cien años de soledad",
    ["datetime"] = "2024-06-30 00:18:23",
    ["highlighted"] = true,
    ["notes"] = "consiguió arrancarle",
    ["page"] = "/body/DocFragment[9]/body/div/div/p[33]/span[11]/text().128",
    ["pos0"] = "/body/DocFragment[9]/body/div/div/p[33]/span[11]/text().128",
    ["pos1"] = "/body/DocFragment[9]/body/div/div/p[33]/span[11]/text().148",
    ["text"] = "● managed to tear him off",
},
```
with key integers increasing sequentially.
Additionally, there is a `"highlight"` table whose entries look as follows:
``` lua
[9] = {
    [1] = {
        ["chapter"] = "Cien años de soledad",
        ["datetime"] = "2024-05-17 23:13:11",
        ["drawer"] = "lighten",
        ["pos0"] = "/body/DocFragment[8]/body/div/div/p[1]/span[2]/text().30",
        ["pos1"] = "/body/DocFragment[8]/body/div/div/p[1]/span[2]/text().53",
        ["text"] = "pelotón de fusilamiento",
    },
    [2] = {
        ["chapter"] = "Cien años de soledad",
        ["datetime"] = "2024-05-17 23:15:03",
        ["drawer"] = "lighten",
        ["pos0"] = "/body/DocFragment[8]/body/div/div/p[1]/span[3]/text().21",
        ["pos1"] = "/body/DocFragment[8]/body/div/div/p[1]/span[3]/text().30",
        ["text"] = "una aldea",
    },
},
```
Here there are keys that start at 9 (presumably arbitrary) and increase sequentially -- this is puzzling to me. Also, the values of this table are also key-value pairs, with keys starting at 1 and increasing sequentially. What is the purpose of this nested structure (as opposed to having just one index as in the case of `"bookmarks"`), and how are highlights split up between first indices (i.e. which highlight tables belong to `[9]`, which to `[10]`, etc.)?

These both contain the starting and ending position of the highlight/bookmark. However, when these refer to `DocFragment[8]` and `DocFragment[9]` I find them in `part03.xhtml` and `part04.xtml` respectively of my ebook. Why the offset? Finally, it appears that `p[m]` denotes paragraph `m` in the `xhtml` file, `span[n]` denotes the `n`th sentence, and `text().d` denotes the `d`th character from the start of the sentence. However, I have run into contradictions with this last rule:

In the `"bookmarks"` table I have an entry:
``` lua
[585] = {
    ["chapter"] = "Cien años de soledad",
    ["datetime"] = "2024-05-19 01:02:27",
    ["highlighted"] = true,
    ["notes"] = "calabazo",
    ["page"] = "/body/DocFragment[8]/body/div/div/p[1]/span[19]/text().109",
    ["pos0"] = "/body/DocFragment[8]/body/div/div/p[1]/span[19]/text().109",
    ["pos1"] = "/body/DocFragment[8]/body/div/div/p[1]/span[19]/text().117",
    ["text"] = "● pumpkin",
},
```
With the corresponding sentence:
``` html
Lo único que logró desenterrar fue una armadura del siglo&nbsp;<small>XV</small> con todas sus partes soldadas por un cascote de óxido, cuyo interior tenía la resonancia hueca de un enorme calabazo lleno de piedras.
```
Here "calabazo" starts more than 109 characters away from the beginning, even if I delete the added html (which I assume one is not supposed to do). Why this discrepancy?

Any help is appreciated!

**Comment by hius07** [comment]
You are describing the old format of highlights/bookmarks.
In the current nightly build and in the next release it is changed to `annotations`.
https://github.com/koreader/koreader/blob/4f75e4163626314bc4fc1b3193b7a5b6f026243d/frontend/apps/reader/modules/readerannotation.lua#L48-L62

**Comment by poire-z** [comment]
> These both contain the starting and ending position of the highlight/bookmark. However, when these refer to `DocFragment[8]` and `DocFragment[9]` I find them in `part03.xhtml` and `part04.xtml` respectively of my ebook. Why the offset?

Because there are 7 other something.xhtml before your part03.xhtml in the book. Look in the file with suffix .opf in the .epub, the section `<spine>` states their ordering.

> Finally, it appears that `p[m]` denotes paragraph `m` in the `xhtml` file, `span[n]` denotes the `n`th sentence, and `text().d` denotes the `d`th character from the start of the sentence.

Not really.
The code and these xpointers (that's how these strings are called) have no notion of paragraph or sentence.
They just work with node tags and text nodes.
`/body/DocFragment[8]/body/div/div/p[1]/span[19]/text().109
`p{1] is the 1st `<p>` child of the div child of the div child of the body child of your 8th xhtml html file.`
span[19]` is the 19th `<span>` in that P. it contains some text node (with possibly some more children and other text nodes), which may only be part of a sentence. The character referenced is the 109th char in that text node.

It's really hard to build them from the outside, it's really something private/undocumented to crengine.
But it's deterministic, so with some courage, you could, after parsing the EPUB, be able to build the exact same xpointers.

**Reply by lucasmyers97** [reply]
Thank you, this is extremely helpful. At risk of asking too much, is there some way that I can see the output of how crengine is parsing my ebook? Is there a command line utility, perhaps, or a lua port? I found the crengine GitHub page difficult to parse (no pun intended).

**Reply by poire-z** [reply]
Not sure what you want.
You could get the single DOM built from the book outputed in crash.log by uncommenting that line:
https://github.com/koreader/koreader/blob/bbe4ce466b1a37fc3eb0fa6b38b4bcc791e58784/frontend/apps/reader/readerui.lua#L327-L328
I usually just use if with small sample book for debugging. On a real big book, you'll get something big, that you could manually trim. But I'm not sure you'll get anything interesting to achieve your aim (your "context" in your first post).
